src/librbd/AioImageRequest.cc

```diff
@@ -724,10 +732,140 @@ void AioImageFlush<I>::send_image_cache_request() {
+    m_off += extent_left;
+  }
+}
+
+template <typename I>
+uint64_t AioImageWriteSame<I>::append_journal_event(
+    const AioObjectRequests &requests, bool synchronous) {
+  I &image_ctx = this->m_image_ctx;
+
+  bufferlist bl;
+  bl.append(m_buf, m_data_len);
+  uint64_t tid;
+  assert(!this->m_image_extents.empty());
+  journal::EventEntry event_entry(journal::AioWriteSameEvent(offset, length,
+							     bl));
+  tid = image_ctx.journal->append_io_event(std::move(event_entry), requests,
+						offset, length, synchronous);
+
+  if (image_ctx.object_cacher == NULL) {
+    AioCompletion *aio_comp = this->m_aio_comp;
+    aio_comp->associate_journal_event(tid);
+  }
+  return tid;
+}
+
+template <typename I>
+void AioImageWriteSame<I>::send_image_cache_request() {
+  I &image_ctx = this->m_image_ctx;
+  assert(image_ctx.image_cache != nullptr);
+
+  AioCompletion *aio_comp = this->m_aio_comp;
+  aio_comp->set_request_count(1);
+  C_AioRequest *req_comp = new C_AioRequest(aio_comp);
+  image_ctx.image_cache->aio_writesame(offset, length, m_buf, m_data_len,
+					m_op_flags, req_comp);
+}
+
+template <typename I>
+void AioImageWriteSame<I>::send_object_cache_requests(const ObjectExtents &object_extents,
+                                                        uint64_t journal_tid) {
+  I &image_ctx = this->m_image_ctx;
+  for (auto p = object_extents.begin(); p != object_extents.end(); ++p) {
+    const ObjectExtent &object_extent = *p;
+
+    bufferlist bl;
+    assemble_extent(object_extent, &bl);
+
+    if (m_off == 0 && (object_extent.length % m_data_len == 0)) {
+      for (int i = 0; i < object_extent.length / m_data_len; i++)
+        bl.append(m_buf, m_data_len);
+    }
+
+    AioCompletion *aio_comp = this->m_aio_comp;
+    C_AioRequest *req_comp = new C_AioRequest(aio_comp);
+    image_ctx.write_to_cache(object_extent.oid, bl, object_extent.length,
+                               object_extent.offset, req_comp, m_op_flags,
+                               journal_tid);
+  }
+  assert(m_off == 0);
+}
+
+template <typename I>
+void AioImageWriteSame<I>::send_object_requests(
+    const ObjectExtents &object_extents, const ::SnapContext &snapc,
+    AioObjectRequests *aio_object_requests) {
+  I &image_ctx = this->m_image_ctx;
+  // cache handles creating object requests during writeback
+  if (image_ctx.object_cacher == NULL) {
+    AbstractAioImageWrite<I>::send_object_requests(object_extents, snapc,
+                                                aio_object_requests);
+  }
+  assert(m_off == 0);
+}
+
+template <typename I>
+AioObjectRequestHandle *AioImageWriteSame<I>::create_object_request(
+    const ObjectExtent &object_extent, const ::SnapContext &snapc,
+    Context *on_finish) {
+  I &image_ctx = this->m_image_ctx;
+  assert(image_ctx.object_cacher == NULL);
+
+  bufferlist bl;
+  assemble_extent(object_extent, &bl);
+
+  AioObjectRequest<I> *req;
+  if (m_off == 0 && (object_extent.length % m_data_len == 0)) {
```

Comment:
Sorry -- incorrect comment when I was attempting to refer to the buffer extents within an object extent